src/std_extensions/collections.rs

```diff
@@ -31,13 +31,18 @@ pub const EXTENSION_NAME: ExtensionId = ExtensionId::new_unchecked("Collections"
 
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 /// Dynamically sized list of values, all of the same type.
-pub struct ListValue(Vec<Value>);
+pub struct ListValue(Vec<Value>, Type);
 
 impl ListValue {
-    /// Create a new [CustomConst] for a list of values.
-    /// (The caller will need these to all be of the same type, but that is not checked here.)
-    pub fn new(contents: Vec<Value>) -> Self {
-        Self(contents)
+    /// Create a new [CustomConst] for a list of values of type `typ`.
+    /// That all values ore of type `typ` is not checked here.
+    pub fn new(typ: Type, contents: Vec<Value>) -> Self {
```

Comment:
Good point that `Value`s don't allow to deduce the type, so we can't really do anything else here. A slight tweak would be to `contents: impl IntoInterator<Value>` allowing callers to omit the `vec!`.

Being slightly wilder, one could have a version taking no Type but a Vec<Const>, but I'm happy to skip that ;-)